sogs/routes/auth.py

```diff
@@ -0,0 +1,236 @@
+    """
+
+    # If we already have a g.user then we are probably a subrequest and want to preserve it, unless
+    # user_reauth has been specifically set (from sogs.routes.subrequest).
+    if hasattr(g, 'user') and not g.user_reauth:
+        return
+
+    g.user_reauth = False
+
+    pk, nonce, ts_str, hash_in = (
+        request.headers.get(f"X-SOGS-{h}") for h in ('Pubkey', 'Nonce', 'Timestamp', 'Hash')
+    )
+
+    missing = sum(x is None or x == '' for x in (pk, nonce, ts_str, hash_in))
+    # If all are missing then we have no user
+    if missing == 4:
+        g.user = None
+        return
+
+    if missing:
+        abort_with_reason(
+            http.BAD_REQUEST,
+            f"Invalid authentication headers: missing {missing}/4 required X-SOGS-* headers",
+        )
+
+    # Parameter input validation
+
+    # TODO: accept/support blinded keys with some other prefix (maybe "15" or "bb" or "55"?)
+    if len(pk) != 66 or pk[0:2] != "05" or not all(x in string.hexdigits for x in pk):
+        abort_with_reason(
+            http.BAD_REQUEST, "Invalid authentication: X-SOGS-Pubkey is not a valid 66-hex digit id"
+        )
+    A = bytes.fromhex(pk[2:])
+
+    try:
+        nonce = utils.decode_hex_or_b64(nonce, 16)
+    except Exception:
+        abort_with_reason(
+            http.BAD_REQUEST,
+            "Invalid authentication: X-SOGS-Nonce must be 16 bytes (encoded as base64 or hex)",
+        )
+
+    try:
+        hash_in = utils.decode_hex_or_b64(hash_in, 64)
+    except Exception:
+        abort_with_reason(
+            http.BAD_REQUEST,
+            "Invalid authentication: X-SOGS-Signature is not base64(86) or hex(128)",
+        )
+
+    try:
+        ts = int(ts_str)
+    except Exception:
+        abort_with_reason(
+            http.BAD_REQUEST, "Invalid authentication: X-SOGS-Timestamp is not a valid timestamp"
+        )
+
+    # Parameter value validation
+
+    now = time.time()
+    if not now - 24 * 60 * 60 <= ts <= now + 24 * 60 * 60:
+        abort_with_reason(
+            http.TOO_EARLY, "Invalid authentication: X-SOGS-Timestamp is too far from current time"
+        )
+
+    user = User(session_id=pk, autovivify=True, touch=False)
+
+    try:
+        query('INSERT INTO user_request_nonces ("user", nonce) VALUES (:u, :n)', u=user.id, n=nonce)
+    except sqlalchemy.exc.IntegrityError:
+        abort_with_reason(http.TOO_EARLY, "Invalid authentication: X-SOGS-Nonce cannot be reused")
+
+    # Signature validation
+
+    # shared_key is hash of a*B || A || B = b*A || A || B where b/B is the server keypair and A is
+    # the session id pubkey.
+    shared_key = nacl.hash.blake2b(
+        crypto_scalarmult(crypto._privkey.encode(), A) + A + crypto.server_pubkey_bytes,
+        digest_size=64,
+        salt=nonce,
+        person=b'sogs.shared_key',
+        encoder=RawEncoder,
+    )
+
+    hasher = nacl.hashlib.blake2b(
+        request.method.encode() + request.path.encode(),
+        digest_size=64,
```

Comment:
make this 32 bytes instead so the final digest is a bit more palatable

Reply:
Made it 42 instead (so that it's distinct from common 32-byte values like pubkeys, b64-encodes without padding, and isn't so long).